From 03c51e5227abd59a9cbe893d953fb154990c5c09 Mon Sep 17 00:00:00 2001
From: Wolfgang Grandegger <wg@aries-embedded.de>
Date: Fri, 28 Apr 2023 15:47:55 +0200
Subject: [PATCH 6/7] board: aries: mcvevk: init the Renesas 5P35023 clock
 generator

The new MCV modules with hw rev. 4.1 use the Renesas 5P35023 clock
generator, which needs to be initialized at startup via I2C bus at
address 0x68. Old modules are still supported; no I2C device will
be found at that address.

Signed-off-by: Wolfgang Grandegger <wg@aries-embedded.de>
---
 board/aries/mcvevk/socfpga.c     | 48 ++++++++++++++++++++++++++++++++
 configs/socfpga_mcvevk_defconfig |  1 +
 2 files changed, 49 insertions(+)

diff --git a/board/aries/mcvevk/socfpga.c b/board/aries/mcvevk/socfpga.c
index f173bf84ac..72544cfc56 100644
--- a/board/aries/mcvevk/socfpga.c
+++ b/board/aries/mcvevk/socfpga.c
@@ -3,3 +3,51 @@
  * Copyright (C) 2015 Marek Vasut <marex@denx.de>
  */
 #include <common.h>
+#include <i2c.h>
+
+#define MCVEVK_SP35023_I2C_BUS  0
+#define MCVEVK_SP35023_I2C_ADDR 0x68
+
+struct sp35023_regs {
+	unsigned char addr;
+	unsigned char val;
+};
+
+static struct sp35023_regs mcvevk_sp35023_regs[] = {
+	{0x1b,0x37},
+	{0x21,0xc0},
+	{0x24,0x8f},
+	{0x1f,0xc7},
+	{0x00,0x08},
+	{0x0f,0x00},
+	{0x23,0x41},
+};
+
+int misc_init_r(void)
+{
+	struct udevice *sp35023 = NULL;
+	int i, err;
+
+	/*
+	 * Configure the Renesas 5P35023 Clock Generator available
+	 * on hw revision >= 4.1
+	 */
+	err = i2c_get_chip_for_busnum(MCVEVK_SP35023_I2C_BUS,
+				      MCVEVK_SP35023_I2C_ADDR, 1, &sp35023);
+	if (err)
+		return 0;
+
+	for (i = 0; i < ARRAY_SIZE(mcvevk_sp35023_regs); i++) {
+		unsigned char addr = mcvevk_sp35023_regs[i].addr;
+		unsigned char val = mcvevk_sp35023_regs[i].val;
+
+		err = dm_i2c_write(sp35023, addr, &val, sizeof(val));
+		if (err) {
+			printf("couldn't write %#x to sp35023 reg at %#x\n",
+			       val, addr);
+			return 1;
+		}
+	}
+
+	return 0;
+}
diff --git a/configs/socfpga_mcvevk_defconfig b/configs/socfpga_mcvevk_defconfig
index 605a006051..8c939614f1 100644
--- a/configs/socfpga_mcvevk_defconfig
+++ b/configs/socfpga_mcvevk_defconfig
@@ -26,6 +26,7 @@ CONFIG_SYS_CONSOLE_ENV_OVERWRITE=y
 # CONFIG_DISPLAY_BOARDINFO is not set
 CONFIG_DISPLAY_BOARDINFO_LATE=y
 CONFIG_CLOCKS=y
+CONFIG_MISC_INIT_R=y
 CONFIG_SPL_PAD_TO=0x10000
 CONFIG_SPL_NO_BSS_LIMIT=y
 # CONFIG_SPL_SHARES_INIT_SP_ADDR is not set
-- 
2.34.1

